core/src/main/java/com/linecorp/armeria/server/cors/CorsPolicy.java

```diff
@@ -0,0 +1,300 @@
+
+import com.linecorp.armeria.common.DefaultHttpHeaders;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.server.cors.CorsConfig.ConstantValueSupplier;
+import com.linecorp.armeria.server.cors.CorsConfig.InstantValueSupplier;
+
+import io.netty.util.AsciiString;
+
+/**
+ * Contains information of the CORS policy with the specified origins.
+ */
+public final class CorsPolicy {
+
+    private static final Logger logger = LoggerFactory.getLogger(CorsPolicy.class);
+    private static final String DELIMITER = ",";
+    private static final Joiner HEADER_JOINER = Joiner.on(DELIMITER);
+    private final Set<String> origins;
+    private final boolean credentialsAllowed;
+    private final boolean nullOriginAllowed;
+    private final long maxAge;
+    private final Set<AsciiString> exposedHeaders;
+    private final Set<HttpMethod> allowedRequestMethods;
+    private final Set<AsciiString> allowedRequestHeaders;
+    private final String joinedExposedHeaders;
+    private final String joinedAllowedRequestHeaders;
+    private final String joinedAllowedRequestMethods;
+    private final HttpHeaders preflightResponseHeaders;
+
+    CorsPolicy(final Set<String> origins, boolean credentialsAllowed, long maxAge,
+               boolean nullOriginAllowed, Set<AsciiString> exposedHeaders,
+               Set<AsciiString> allowedRequestHeaders, Set<HttpMethod> allowedRequestMethods,
+               boolean preflightResponseHeadersDisabled,
+               Map<AsciiString, Supplier<?>> preflightResponseHeaders) {
+        this.origins = origins;
+        this.credentialsAllowed = credentialsAllowed;
+        this.maxAge = maxAge;
+        this.nullOriginAllowed = nullOriginAllowed;
+        this.exposedHeaders = ImmutableSet.copyOf(exposedHeaders);
+        this.allowedRequestMethods = EnumSet.copyOf(allowedRequestMethods);
+        this.allowedRequestHeaders = ImmutableSet.copyOf(allowedRequestHeaders);
+        joinedExposedHeaders = HEADER_JOINER.join(this.exposedHeaders);
+        joinedAllowedRequestMethods = this.allowedRequestMethods
+                .stream().map(HttpMethod::name).collect(Collectors.joining(DELIMITER));
+        joinedAllowedRequestHeaders = HEADER_JOINER.join(this.allowedRequestHeaders);
+        final Map<AsciiString, Supplier<?>> preflightResponseHeadersMap;
+        if (preflightResponseHeadersDisabled) {
+            preflightResponseHeadersMap = Collections.emptyMap();
+        } else if (preflightResponseHeaders.isEmpty()) {
+            preflightResponseHeadersMap = ImmutableMap.of(
+                    HttpHeaderNames.DATE, InstantValueSupplier.INSTANCE,
+                    HttpHeaderNames.CONTENT_LENGTH, ConstantValueSupplier.ZERO);
+        } else {
+            preflightResponseHeadersMap = preflightResponseHeaders;
+        }
+        final HttpHeaders preflightHeaders;
```

Comment:
I realized preflight headers cannot be cached unless it's disabled. For example, `InstantValueSupplier` above will have to produce a different timestamp for each request. We could cache it only when:

- preflight headers are disabled or empty, or
- all suppliers are `ConstantValueSupplier`

(or we could just give up caching preflight headers?)